examples/rhat_monitor.cpp

```diff
@@ -0,0 +1,387 @@
+      sample_.append_draw(logp, theta);
+      q_.emplace(logp_stats_.sample_stats());
+      if (st.stop_requested()) {
+        break;
+      }
+    }
+  }
+
+  const Sample& sample() const { return sample_; }
+
+  Sample&& take_sample() { return std::move(sample_); }
+
+  double mean_logp() const { return logp_stats_.mean(); }
+
+  double sample_variance_logp() const { return logp_stats_.sample_variance(); }
+
+  std::size_t count_logp() const { return logp_stats_.count(); }
+
+  const WelfordAccumulator& logp_stats() const { return logp_stats_; }
+
+ private:
+  std::size_t chain_id_;
+  std::size_t draws_per_chain_;
+  Sampler& sampler_;
+  Sample sample_;
+  WelfordAccumulator logp_stats_;
+  Queue& q_;
+  std::latch& start_gate_;
+};
+
+static void controller_loop(std::stop_token st, std::vector<Queue>& queues,
+                            std::vector<std::jthread>& workers,
+                            double rhat_threshold, std::latch& start_gate) {
+  start_gate.wait();
+  const std::size_t M = static_cast<std::size_t>(queues.size());
+  std::vector<double> chain_means(M, std::numeric_limits<double>::quiet_NaN());
+  std::vector<double> chain_variances(M,
+                                      std::numeric_limits<double>::quiet_NaN());
+  while (true) {
+    for (std::size_t m = 0; m < M; ++m) {
+      SampleStats u;
+      while (queues[m].pop(u)) {
+        chain_means[m] = u.sample_mean;
+        chain_variances[m] = u.sample_var;
+      }
+    }
+    double variance_of_means = variance(chain_means);
+    double mean_of_variances = mean(chain_variances);
+    double r_hat = std::sqrt(1 + variance_of_means / mean_of_variances);
+    // print is just for debugging/demonstrating runs
+    std::cout << "variance of means=" << variance_of_means
+              << "; mean of variances=" << mean_of_variances
+              << "; r_hat = " << r_hat << '\n';
+    if (r_hat <= rhat_threshold) {
+      for (auto& w : workers) {
+        w.request_stop();
+      }
+      break;
+    }
+  }
+}
+
+template <typename Sampler>
+std::vector<Sample> sample(std::vector<Sampler>& samplers,
+                           double rhat_threshold,
+                           std::size_t max_draws_per_chain) {
+  using namespace std::chrono_literals;
+  using Task = ChainTask<Sampler>;
+
+  const std::size_t M = samplers.size();
+  std::vector<Queue> queues(M);
+  std::latch start_gate(M);
+  std::vector<Task> tasks;
+  tasks.reserve(M);
+  for (std::size_t m = 0; m < M; ++m) {
+    tasks.emplace_back(m, max_draws_per_chain, samplers[m], queues[m],
+                       start_gate);
+  }
+  {  // scope for jthread RAII control
+    std::vector<std::jthread> workers;
+    workers.reserve(M);
+    for (std::size_t m = 0; m < M; ++m) {
+      workers.emplace_back(std::ref(tasks[m]));
+    }
+    std::jthread controller([&](std::stop_token st) {
+      interactive_qos();
+      controller_loop(st, queues, workers, rhat_threshold, start_gate);
+    });
```

Comment:
is there a reason to run the controller thread on its own thread rather than on the main thread here?